### ISSUE TITLE
allow workaround for counter-intuitive contexts behavior keeping backward compatibility

### ISSUE THREAD

**Issue by mzatko** [liquibase/liquibase#3458]
## Description
changeSet with context specified applies even when update executed with no --contexts specified at runtime
I assume decision about this behavior was made pretty long time ago, so you want to keep backward compatibility. However, googling all around, many people found this behavior as counter-intuitive.

Actual problem is the need to introduce behavior like:
1. if there is no --contexts specified at runtime, only changesets without any context specified will be applied
2. if there is(are) some --contexts specified at runtime, all changesets without any context specified will be applied and all changesets with matching context expression specified will be updated.

Would it be possible to add some pseudo-context like **nocontexts** (or similar) to be available for checks when no --contexts specified at runtime? 
So we'll be able to add a check to changeset's/changelog's context expression.
something like `context="!nocontexts AND specificcontext"`
Maybe in case of "the opposite approach"  pseudo-context's name could be **notblank** (or similar) and condition then
`context="notblank AND specificcontext"`

Another possibility would be to add another attribute to changeSet/changeLog, something like strictContext=true/false, with default value "false"

**Comment by nvoxland** [comment]
Yes, the current behavior does trip people up at times. 

The rationale behind it is this: when first running liquibase you have no contexts defined and just run `liquibase update` and everything runs. As you start to use contexts and add `contextFilter` to changesets what should the behavior be for times when you just run `liquibase update` like before? Still run all changesets? Or only run ones with no contextFilter added? We kept with the "still" in "still run" for consistency. 

But in the end, some people want it to be "still run all" and some people want "only run ones without contextFilter" and if we change that default behavior it's going to be breaking for people. 

But, we can certainly add new settings to make that behavior modifiable. 

We could make a global config setting like `filter-unspecfied-context=true|false`  (I'm sure there is a better name) which can toggle the match logic for when context is not set so a simple `liquibase update` will not run any changeSets with contextFilter="whatever". 

Alternately, there is the suggestion of having a `nocontexts` or something default value for --contexts like you suggest. That is interesting in that it allows you to have contextFilters that explicitly handle the case of not specifying a context. I wonder if that will be more complex to implement though? We have the CONTEXTS_ARG settings which don't have a default value, and we probably can't currently have a default value that is itself a configuration value. But, we could maybe still add that support into the filter matching logic? In this case, the new global setting could be something different than `filter-unspecfied-context` like `default-context=nocontext` or whatever the user wants? 

That could also maybe just be a feature of  `filter-unspecfied-context=true` where we enable the filtering and have a hard-coded `nocontext` context that is made available? I'd maybe think this is the best solution, but will probably take some playing with the code and with examples to see.

If anyone wants to take a stab at implementing it and sending a PR, we're always open to that.

**Comment by filipelautert** [comment]
We're currently triaging our backlog to ensure we can focus on the issues that matter most. With a large number of open issues, it becomes difficult to give each one the attention it deserves.

This issue has been closed due to inactivity — there have been no updates for over 30 days. If this is still relevant to you, please feel free to reopen it or create a new issue with updated context. We appreciate your contribution!

**Comment by mzatko** [comment]
I think we need to be able to deal with this counter-intuitive behavior somehow so the issue needs to be re-opened

**Comment by filipelautert** [comment]
Whow, that one is old. Thanks, moving this one!

**Comment by filipelautert** [comment]
Looks like copilot has some idea on how to fix this..... https://github.com/liquibase/liquibase/pull/7650

**Comment by mzatko** [comment]
yeah, #7650 looks good to me, it is exactly what I suggested ~4years ago

**Comment by filipelautert** [comment]
It took 4 years for AI to ` take a stab at implementing it and sending a PR` :)
